constraint/pkg/client/e2e_test.go

```diff
@@ -1,465 +1,453 @@
-	// specifies a NamespaceSelector, but the client does not have any Namespaces
-	// cached. Thus, it is impossible for this Constraint to run properly as it
-	// cannot determine if the object's Namespace matches the selector or not.
-	//
-	// This differs from a normal "DENIED" as we were unable to even run the
-	// Constraint.
-	for _, tc := range denyAllCases {
-		t.Run(tc.name, func(t *testing.T) {
-			ctx := context.Background()
+	want := []*types.Result{{
+		Msg:               "got qux but want bar for data",
+		EnforcementAction: "deny",
+		Constraint: clienttest.MakeConstraint(t, clienttest.KindCheckData, "constraint",
+			clienttest.WantData("bar")),
+		Metadata: map[string]interface{}{"details": map[string]interface{}{"got": "qux"}},
+	}}
 
-			c, err := newTestClient()
-			if err != nil {
-				t.Fatal(err)
-			}
+	results := responses.Results()
 
-			_, err = c.AddTemplate(newConstraintTemplate("Foo", denyTemplateRego))
-			if err != nil {
-				t.Fatalf("got AddTemplate: %v", err)
-			}
-			goodNamespaceSelectorConstraint := `
-{
-	"apiVersion": "constraints.gatekeeper.sh/v1alpha1",
-	"kind": "Foo",
-	"metadata": {
-  	"name": "foo-pod"
-	},
-	"spec": {
-  	"match": {
-    	"kinds": [
-      	{
-			"apiGroups": [""],
-        	"kinds": ["Pod"]
-		}],
-		"namespaceSelector": {
-			"matchExpressions": [{
-	     		"key": "someKey",
-				"operator": "Blah",
-				"values": ["some value"]
-			}]
-		}
-	},
-  	"parameters": {
-    	"key": ["value"]
-		}
+	diffOpt := cmpopts.IgnoreFields(types.Result{}, "Review", "Resource")
+	if diff := cmp.Diff(want, results, diffOpt); diff != "" {
+		t.Error(diff)
 	}
 }
-`
-			u := &unstructured.Unstructured{}
-			err = json.Unmarshal([]byte(goodNamespaceSelectorConstraint), u)
-			if err != nil {
-				t.Fatalf("got Unable to parse constraint JSON: %v", err)
-			}
 
-			if _, err := c.AddConstraint(ctx, u); err != nil {
-				t.Fatalf("got AddConstraint: %v", err)
-			}
-
-			rsps, err := c.Review(ctx, targetData{Name: "Sara", ForConstraint: "Foo"})
-			if err != nil {
-				t.Fatalf("got Review: %v", err)
-			}
-
-			got := rsps.Results()
-			want := []*types.Result{{
-				Constraint:        u,
-				EnforcementAction: "deny",
-				Msg:               denied,
-			}, {
-				Constraint:        u,
-				EnforcementAction: "deny",
-				Msg:               rejection,
-			}}
-
-			if diff := cmp.Diff(want, got, cmpopts.IgnoreFields(types.Result{},
-				"Metadata", "Review", "Resource")); diff != "" {
-				t.Error(diff)
-			}
-		})
+func TestClient_Audit(t *testing.T) {
+	tests := []struct {
+		name        string
+		templates   []*templates.ConstraintTemplate
+		constraints []*unstructured.Unstructured
+		objects     []*clienttest.Object
+		want        []*types.Result
+	}{
+		{
+			name:        "empty client returns empty audit",
+			templates:   nil,
+			constraints: nil,
+			objects:     nil,
+		},
+		{
+			name: "no template returns empty audit",
```

Comment:
nit: "no constraint returns empty audit"?

Reply:
Good catch. Also added case for empty templates but with object.